### DISCUSSION TITLE
Not all files upload to GitHub

### DISCUSSION THREAD

**Discussion by matbishop** [community/community#45219] [Other Feature Feedback, Questions, & Ideas] [answered]
### Select Topic Area

Question

### Body

Hi all. Apologies in advance for my (perhaps trivial) questions as I am a new user.

I have been struggling to get to grips with Git for quite a while - not necessarily using it but rather setup and general use.

I have two questions at the moment: Firstly, not all my files are being uploaded to GitHub for a reason I do not know. I just installed GitHub Desktop to see if it can make my life easier. I deleted all my repos online and tried adding my existing files to a new repo created in the app. The new repo shows online with my files, but there are certain files that are simply missing, and I cannot determine why.

I have three folders in this particular folder, but only one is fully uploaded online (CS50x); one of them is **not uploaded at all** (CS50p) and the third one is only **partially uploaded** (CS50w). The CS50w folder has three subfolders, but as can be seen in the screenshot only one is uploaded - project0. 

This also brings me to my second question - why does GitHub display some folders differently? In the screenshot CS50x is displayed differently to CS50w - is this because CS50w only contains one folder? (Although in reality it does not)

Any help would be sincerely appreciated.

Kind regards,
Matthew 

Screenshot:
<img width="1253" alt="Screenshot 2023-01-24" src="https://user-images.githubusercontent.com/108872176/214306726-dc4df784-fde7-48f2-9574-f4574d5bd10d.png">

**Comment by VladCuciureanu** [comment]
Hey Matthew!

Regarding the files that are not being uploaded... By 'files' do you mean empty folders?
I'm asking this because empty folders _are not_ picked up by git by default.
If you _need_ to upload the empty folders and don't have any files to add yet, adding an empty '.gitkeep' or '.keep' file inside is one way to do it.

Regarding the different ways GitHub displays folders, what you're seeing here is 'compact paths'. Let me expand on this...

When GitHub detects the following path pattern:
<img width="112" alt="Screenshot 2023-01-25 at 19 14 57" src="https://user-images.githubusercontent.com/71708164/214633705-3abd234e-87dc-44c8-bfbc-bf93d78b125e.png">
It rids you of clicking multiple times through multiple levels of folders by adding a shortcut to the `child` folder:
`grandparent/parent/child`
This would not be possible in the following case:
<img width="114" alt="Screenshot 2023-01-25 at 19 16 09" src="https://user-images.githubusercontent.com/71708164/214634127-4a7d0c8f-ae5d-4c69-ad33-730e9b50e8bc.png">
That's because `grandparent` now contains _two_ files and GitHub can't predict which one you'll access.

In your case, from what I can tell from your screenshot:
- 'CS50w' only contains 'project0'.
- 'project0' only contains 'search'.
- 'CS50x' contains multiple files.

... or at least I predict that's the case on your remote repo.

Hope this was of help ^^
Vlad Cuciureanu

**Reply by matbishop** [reply]
Hi Vlad,

Thank you for taking the time to help me, I appreciate it!

Thanks for the folder structure explanation. Unfortunately the files I am referring to are **not** empty folders - they very much have files in them. 'CS50w' contains many projects (with files in them - not empty folders). Correct - 'project0' only contains 'search'. Lastly the folder CS50p (also which contains many projects) is not uploaded at all.

I wondered if these files were being left out purposefully because of gitignore or something else (not that I know how to configure this), but I don't know as like I said I tried creating new repositories, deleteing git files and uploading from scratch. Still the files do not show.

Kind regards,
Matthew

**Comment by matbishop** [comment, marked as answer]
I figured out the problem to this and though I might as well share it here.

I used a virtual environment for the projects, and when creating a virtual environment using `virtualenv .` a gitignore file is automatically created that excludes all files. This will exclude the entire project if it is in the root directory and not a separate directory for just the virtual environment files. The gitignore file on Mac is typically hidden and can be shown with the keyboard shortcut Cmd + Shift + .
